### PR TITLE
Use opaque pointers for rct_window

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openrct2/world/TileInspector.cpp [hunk 1]
```diff
@@ -134,15 +134,15 @@ sint32 tile_inspector_insert_corrupt_at(sint32 x, sint32 y, sint16 elementIndex,
             windowTileInspectorElementCount++;
 
             // Keep other elements (that are not being hidden) selected
-            if (tileInspectorWindow->selected_list_item > elementIndex)
-            {
-                tileInspectorWindow->selected_list_item++;
-            }
+//            if (tileInspectorWindow->selected_list_item > elementIndex)
```
Comment (marijnvdwerf) [inline src/openrct2/world/TileInspector.cpp]: `selected_list_item`

File: src/openrct2/Editor.cpp [hunk 2]
```diff
@@ -442,11 +442,11 @@ namespace Editor
...
         rct_window   * w        = window_get_main();
-        rct_viewport * viewport = w->viewport;
+        rct_viewport * viewport = window_get_viewport(w);
 
-        w->viewport_target_sprite = SPRITE_INDEX_NULL;
-        w->saved_view_x           = gSavedViewX;
-        w->saved_view_y           = gSavedViewY;
+//        w->viewport_target_sprite = SPRITE_INDEX_NULL;
+//        w->saved_view_x           = gSavedViewX;
+//        w->saved_view_y           = gSavedViewY;
```
Comment (marijnvdwerf) [inline src/openrct2/Editor.cpp]: `viewport_target_sprite`, `saved_view_x`, `saved_view_y`

File: src/openrct2/actions/GuestSetNameAction.hpp [hunk 3]
```diff
@@ -121,7 +121,7 @@ struct GuestSetNameAction : public GameActionBase<GAME_COMMAND_SET_GUEST_NAME, G
         rct_window * w = window_find_by_class(WC_GUEST_LIST);
         if (w != NULL)
         {
-            w->no_list_items = 0;
+//            w->no_list_items = 0;
```
Comment (marijnvdwerf) [inline src/openrct2/actions/GuestSetNameAction.hpp]: `no_list_items`

File: src/openrct2/interface/Theme.cpp [hunk 4]
```diff
@@ -889,15 +889,15 @@ extern "C"
...
+//        for (rct_window *w = g_window_list; w < gWindowNextSlot; w++)
+//        {
+//            colour_scheme_update(w);
+//        }
     }
 
     void colour_scheme_update(rct_window * window)
     {
-        colour_scheme_update_by_class(window, window->classification);
+//        colour_scheme_update_by_class(window, window->classification);
```
Comment (marijnvdwerf) [inline src/openrct2/interface/Theme.cpp]: `classification`

File: src/openrct2/interface/Theme.cpp [hunk 5]
```diff
@@ -922,10 +922,10 @@ extern "C"
         }
 
         for (sint32 i = 0; i < 6; i++) {
-            window->colours[i] = windowTheme->Colours[i];
+//            window->colours[i] = windowTheme->Colours[i];
```
Comment (marijnvdwerf) [inline src/openrct2/interface/Theme.cpp]: `colours`

File: src/openrct2/interface/Theme.cpp [hunk 6]
```diff
@@ -922,10 +922,10 @@ extern "C"
...
         }
         // Some windows need to be transparent even if the colours aren't.
         // There doesn't seem to be any side-effects for all windows being transparent
-        window->flags |= WF_TRANSPARENT;
+//        window->flags |= WF_TRANSPARENT;
```
Comment (marijnvdwerf) [inline src/openrct2/interface/Theme.cpp]: `flags`

File: src/openrct2/ride/Ride.cpp [hunk 7]
```diff
@@ -1004,7 +1004,7 @@ static rct_window *ride_create_or_find_construction_window(sint32 rideIndex)
     rct_window *w;
 
     w = window_find_by_class(WC_RIDE_CONSTRUCTION);
-    if (w == nullptr || w->number != rideIndex) {
+    if (w == nullptr /*|| w->number != rideIndex*/) {
```
Comment (marijnvdwerf) [inline src/openrct2/ride/Ride.cpp]: `number`

File: src/openrct2/ride/Ride.cpp [hunk 8]
```diff
@@ -7352,9 +7352,9 @@ static money32 ride_set_vehicles(uint8 rideIndex, uint8 setting, uint8 value, ui
 
     rct_window *w = window_find_by_number(WC_RIDE, rideIndex);
     if (w != nullptr) {
-        if (w->page == 4) { // WINDOW_RIDE_PAGE_COLOUR
-            w->vehicleIndex = 0;
-        }
+//        if (w->page == 4) { // WINDOW_RIDE_PAGE_COLOUR
```
Comment (marijnvdwerf) [inline src/openrct2/ride/Ride.cpp]: `page`

File: src/openrct2/ride/Ride.cpp [hunk 9]
```diff
@@ -7352,9 +7352,9 @@ static money32 ride_set_vehicles(uint8 rideIndex, uint8 setting, uint8 value, ui
...
+//            w->vehicleIndex = 0;
```
Comment (marijnvdwerf) [inline src/openrct2/ride/Ride.cpp]: `vehicleIndex`

File: src/openrct2/ride/Vehicle.cpp [hunk 10]
```diff
@@ -9870,7 +9870,8 @@ void vehicle_invalidate_window(rct_vehicle * vehicle)
         return;
 
     ride             = get_ride(vehicle->ride);
-    viewVehicleIndex = w->ride.view - 1;
+//    viewVehicleIndex = w->ride.view - 1;
```
Comment (marijnvdwerf) [inline src/openrct2/ride/Vehicle.cpp]: `ride.view`

**Comment by marijnvdwerf** [comment]
I'd like the non-UI files to not need to import `window_private.h`. This means that they shouldn't access window properties directly, and preferably even only broadcast an intent that we only react to on the UI layer.

The ultimate goal is to make the game mostly unaware of the UI layer existing, only broadcasting events if something happens that would cause the UI to update.

Files I'm talking about:
Editor.cpp

**Comment by IntelOrca** [comment]
I am for removing the broadcasting altogether as it assumes the logic knows what windows show. We aim for an extensible system with scripting support so it should be the windows' job to check if it needs to invalidate by polling for state changes.

**Comment by marijnvdwerf** [comment]
How would you expect things like the staff list to update, without events?

**Comment by IntelOrca** [comment]
It would just poll for changes periodically.

**Comment by ZehMatt** [comment]
@IntelOrca that sounds like a really strange data model. All common UI interfaces have some internal message system where you can define your own messages so its up to the window how to deal with it. The polling sounds rather harsh and has a lot of states you need to consider then.

**Pull request merged**